File: src/week1/report2/TicketOffice.java
package week1.report2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TicketOffice {
  private Long amount;
  private Map<Movie,List<Ticket>> movies = new HashMap<>();
  
  public TicketOffice(Long amount){this.amount = amount;}
  
  public Ticket getTicketWithFee(Movie movie){
    if(!movies.containsKey(movie)) return Ticket.EMPTY;
    else if(movies.get(movie).size() == 0) return Ticket.EMPTY;
    else{
      Ticket ticket = movies.get(movie).remove(0);
      amount += ticket.getFee();
      return ticket;
    }
  }
  public Ticket getTicketWithNoFee(Movie movie){
    if(!movies.containsKey(movie)) return Ticket.EMPTY;
    else if(movies.get(movie).size() == 0) return Ticket.EMPTY;
    else{
      return movies.get(movie).remove(0);
    }
  }
  public Long getTicketPrice(Movie movie){
    if(!movies.containsKey(movie)) return 0L;
    else if(movies.get(movie).size() == 0) return 0L;
    else return movies.get(movie).get(0).getFee();
  }
  
  public void addTicket(Movie movie , long num){
    List<Ticket> tickets = new ArrayList<>();
    while(num-- > 0) {
      tickets.add(new Ticket(movie));
    }
    movies.put(movie, tickets);
  }
}

File: Swift/week1/CodeSpitzExample/Example/TicketOffice.swift
import Foundation

class TicketOffice {
    private var amount: Long
    private var tickets = [Ticket]()
            
    init(amount: Long) {
        self.amount = amount
    }
    
    func addTicket(ticket: Ticket) {
        self.tickets.append(ticket)
    }
    
    func getTicketWithFee() -> Ticket {
        if tickets.isEmpty {
            return Ticket.empty
        } else {
            let ticket = tickets.remove(at: 0)
            amount += ticket.fee
            return ticket
        }
    }
    
    // TicketOffice 는 Invitation 에 대해 모른다.
    // VIP, 연간회원등 추가될 정책을 모두 수용할려면 Invitation 에 구체적인 정보를 몰라야 한다.
    func getTicketWithNoFee() -> Ticket {
        if tickets.isEmpty {
            return Ticket.empty
        } else {
            return tickets.remove(at: 0)
        }
    }
    
    func getTicketPrice() -> Long {
        if tickets.isEmpty {
            return 0
        } else {
            return tickets[0].fee
        }
    }
}
